### PR TITLE
blazing fast mapped()

### PR DESCRIPTION
**Pull request by Elkasitu**
Reimplementation of (part of) #38850

TODO:
 - [x] Make `_RelationalMulti.__get__` use the new and improved `mapped` instead of naively reimplementing the old one

### CODE REVIEW

**Comment by fmdl** [comment]
FYI @Elkasitu I have work to make mapped faster : https://github.com/odoo/odoo/pull/30975

**Comment by Elkasitu** [comment]
@fmdl Unfortunately the problem with using `prefetch_fields=False` is that, while you gain in performance locally, you potentially lose performance globally and it's very hard to see / test where and how much you lose.

File: odoo/fields.py [hunk 1]
```diff
@@ -660,6 +660,37 @@ def get(key, get_context=env.context.get):
 
         return tuple(get(key) for key in self.depends_context)
 
+    def mapped(self, records):
```
Comment (rco-odoo) [inline odoo/fields.py]: docstring

Comment (rco-odoo) [reply odoo/fields.py]: Why does `field.mapped()` return a list of cache-formatted values?
I would expect `field.mapped()` to be equivalent to the multi version of `field.__get__()`.

File: odoo/fields.py [hunk 2]
```diff
@@ -660,6 +660,37 @@ def get(key, get_context=env.context.get):
...
+        if self.compute:
+            # a recompute must be forced for records to_recompute that intersect with input records
+            # otherwise an infinite loop will occur on the next call to flush() / recompute()
+            # as the value will be False in cache and will never trigger a MissingError which,
+            # in turn, triggers a recompute
+            to_compute_ids = records.env.all.tocompute.get(self, set())
+            if not to_compute_ids.isdisjoint(records._ids):
+                to_compute_ids = to_compute_ids.intersection(records._ids).difference(
+                    records.env._protected.get(self, ()))
+                self.compute_value(records.browse(to_compute_ids.intersection(records._ids)))
```
Comment (rco-odoo) [inline odoo/fields.py]: Why intersecting again with `records._ids`?
Note also that `to_compute_ids` may be empty once protected ids are subtracted.

Why not doing:
```python
to_compute_ids = records.env.all.tocompute.get(self, ())
to_protect_ids = records.env._protected.get(self, ())
ids = [id_ for id_ in record._ids
       if id_ in to_compute_ids and id_ not in to_protect_ids]
if ids:
    self.compute_value(records.browse(ids))
```

Comment (rco-odoo) [reply odoo/fields.py]: Wait, we can simplify this!  By design, a field cannot be both to compute and protected on a given record.  So you can actually ignore protected fields here...

File: odoo/fields.py [hunk 3]
```diff
@@ -660,6 +660,37 @@ def get(key, get_context=env.context.get):
...
+
+            remaining_new = remaining.filtered(lambda r: not r.id)
+            if remaining_new:
+                # populate the cache for each NewId record
+                [self.__get__(rec, type(rec)) for rec in remaining_new]
+            else:
+                missing = records._browse(records.env, (remaining._ids[0],), tuple(remaining._ids))
+                # trigger the prefetch of all remaining records and store the values in cache
+                self.__get__(missing, type(missing))
+            return vals + self.mapped(remaining)
```
Comment (rco-odoo) [inline odoo/fields.py]: I am not fan of a recursive call for such an obviously iterative process.
What about a simple while loop?
```python
vals = records.env.cache.get_until_miss(records, self)
while len(vals) < len(records):
    remaining = records[len(vals):]
    # simplified, simply trigger prefetching from the first remaining record
    self.__get__(first(remaining), type(remaining))
    vals += records.env.cache.get_until_miss(remaining, self)
```

File: odoo/fields.py [hunk 4]
```diff
@@ -785,6 +816,9 @@ def convert_to_record(self, value, record):
         """
         return False if value is None else value
 
+    def convert_to_record_multi(self, values, record):
+        return [self.convert_to_record(value, record) for value in values]
```
Comment (rco-odoo) [inline odoo/fields.py]: Performance: avoid the method lookup in the loop
```suggestion
        convert = self.convert_to_record
        return [convert(value, record) for value in values]
```

File: odoo/fields.py [hunk 5]
```diff
@@ -2453,6 +2485,11 @@ def convert_to_record(self, value, record):
         prefetch_ids = IterableGenerator(prefetch_many2one_ids, record, self)
         return record.pool[self.comodel_name]._browse(record.env, ids, prefetch_ids)
 
+    def convert_to_record_multi(self, values, record):
+        prefetch_ids = IterableGenerator(prefetch_many2one_ids, record, self)
+        ids = tuple(unique(val for val in values if val is not None))
```
Comment (rco-odoo) [inline odoo/fields.py]:
```suggestion
        ids = tuple(unique(id_ for id_ in values if id_ is not None))
```

File: odoo/fields.py [hunk 6]
```diff
@@ -2722,6 +2759,17 @@ def convert_to_record(self, value, record):
             corecords = corecords.filtered(Comodel._active_name).with_prefetch(prefetch_ids)
         return corecords
 
+    def convert_to_record_multi(self, values, record):
+        prefetch_ids = IterableGenerator(prefetch_x2many_ids, record, self)
+        Comodel = record.pool[self.comodel_name]
+        ids = tuple(unique(i for v in values for i in v))
```
Comment (rco-odoo) [inline odoo/fields.py]:
```suggestion
        ids = tuple(unique(id_ for ids in values for id_ in ids))
```

File: odoo/models.py [hunk 7]
```diff
@@ -5269,6 +5270,8 @@ def filtered(self, func):
         if isinstance(func, str):
             name = func
             func = lambda rec: any(rec.mapped(name))
+            # populate cache
+            self.mapped(name)
```
Comment (rco-odoo) [inline odoo/models.py]: Does this really makes a difference?

File: odoo/fields.py (hunk 2 above)
Comment (Elkasitu) [reply odoo/fields.py]: /!\ Recursive fields must be computed one by one; while the records `to_compute` and `records` intersect, we must compute the first record of the recordset

File: odoo/models.py (hunk 7 above)
Comment (Elkasitu) [reply odoo/models.py]: Yes but maybe I should add a more descriptive comment

File: odoo/fields.py [hunk 8]
```diff
@@ -660,6 +660,47 @@ def get(key, get_context=env.context.get):
...
+            # as the value will be False in cache and will never trigger a MissingError which,
+            # in turn, triggers a recompute
+            to_compute_ids = records.env.all.tocompute.get(self, ())
+            ids = [id_ for id_ in records._ids if id_ in to_compute_ids]
+
+            if self.recursive:
+                # recursive computed fields are a special case and must be computed one by one
+                # in order, as a record down the stack may depend on the computed value of a
+                # previous record
+                ids = ids[:1]
```
Comment (rco-odoo) [inline odoo/fields.py]: This is not correct: you have to force the computation of `self` on *all* records to compute!
```python
for record in records.browse(ids):
    self.__get__(record, type(records))
```
Note that this `for` loop works for all computed fields...  Maybe worth trying its performance.

File: odoo/fields.py [hunk 9]
```diff
@@ -785,6 +826,11 @@ def convert_to_record(self, value, record):
         """
         return False if value is None else value
 
+    def convert_to_record_multi(self, values, record):
```
Comment (rco-odoo) [inline odoo/fields.py]: Make the API more consistent: `record` is not a single record
```suggestion
    def convert_to_record_multi(self, values, records):
```

File: odoo/fields.py [hunk 10]
```diff
@@ -2453,6 +2497,11 @@ def convert_to_record(self, value, record):
         prefetch_ids = IterableGenerator(prefetch_many2one_ids, record, self)
         return record.pool[self.comodel_name]._browse(record.env, ids, prefetch_ids)
 
+    def convert_to_record_multi(self, values, record):
+        prefetch_ids = IterableGenerator(prefetch_many2one_ids, record, self)
+        ids = tuple(unique(id_ for id_ in values if id_ is not None))
+        return record.pool[self.comodel_name]._browse(record.env, ids, prefetch_ids)
```
Comment (rco-odoo) [inline odoo/fields.py]: I am just wondering: is `prefetch_ids` necessary?
```python
records = model.browse(ids)

# each `field.mapped` returns a recordset that will not prefetch beyond itself
records.mapped('partner_id.country_id.name')

for record in records:
    # each `field.mapped` returns a record that could prefetch more than itself
    record.mapped('partner_id.country_id.name')
```
In the first case, we could simplify: `prefetch_ids=ids`...

But I don't know whether it is worth the complexification.

Comment (Elkasitu) [reply odoo/fields.py]: Conclusion: not for now...

**Comment by rco-odoo** [comment]
@robodoo r+

**Pull request closed without merging**